Importing libraries.

In [1]:
# import libraries

import pandas as pd
import numpy as np
import os

Creating data path.

In [2]:
# Create data path
path = r"C:\Users\Jason\Instacart Basket Analysis"

In [3]:
path

'C:\\Users\\Jason\\Instacart Basket Analysis'

Importing dataframe from previous exercise.

1. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [4]:
# Importing df
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_prods_merge_derived.pkl'))

In [5]:
# Renaming df for ease
df = df_ords_prods_merge

In [6]:
# DF shape and head check
df.shape

(32404859, 16)

In [7]:
# DF shape and head check
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                     7.0       14084                  2          0   
2                     7.0       12427                  3          0   
3                     7.0       26088                  4          0   
4                     7.0       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  match    busiest_days busiest_period_of_day  
0  both  Regularly busy        Average orders  
1  both  Regularly busy        Average orders  
2  both  Regularly busy        Average orders  
3  both  Regularly busy        Average orders  
4  both  Regularly busy        Average orders

2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [8]:
# Finding the aggregated mean of the “order_number” column grouped by “department_id” for a subset of my dataframe
df.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

3. Analyze the result. How do the results for the entire dataframe differ from those of the subset?
   - The full dataframe shows a higher average/mean order numbers across the departments.
   - Department 21 is the leader in mean values across both datasets, however, it is noticeably higher in the subset, as opposed to the full dataset.
   - We see some departments that show consistent mean order numbers across both datasets (13 and 17.) This reflects more stable behaviour for the customers, I think.
   - The entire dataframe shows a more comprehensive view of the ordering frequency, but we do see more diverse patterns across departments in the subset.

4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [9]:
# Using the transform() function for data aggregation
df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\Jason\AppData\Local\Temp\ipykernel_10408\423542265.py:2: FutureWarning: The provided callable <function max at 0x0000022CF15B2C00> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)


In [10]:
# Checking output
df.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   
5   2398795        1             2                   3                  7   
6   2398795        1             2                   3                  7   
7   2398795        1             2                   3                  7   
8   2398795        1             2                   3                  7   
9   2398795        1             2                   3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                     7.0       14084                  2          0   
2                     7.0       12427                  3          0   
3                     7.0       26088                  4          0   
4                     7.0       26405                  5          0   
5                    15.0         196                  1          1   
6                    15.0       10258                  2          0   
7                    15.0       12427                  3          1   
8                    15.0       13176                  4          0   
9                    15.0       26088                  5          1   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                     Soda        77              7     9.0   
6                               Pistachios       117             19     3.0   
7                      Original Beef Jerky        23             19     4.4   
8                   Bag of Organic Bananas        24              4    10.3   
9               Aged White Cheddar Popcorn        23             19     4.7   

  match    busiest_days busiest_period_of_day  max_order  
0  both  Regularly busy        Average orders         10  
1  both  Regularly busy        Average orders         10  
2  both  Regularly busy        Average orders         10  
3  both  Regularly busy        Average orders         10  
4  both  Regularly busy        Average orders         10  
5  both    Slowest days        Average orders         10  
6  both    Slowest days        Average orders         10  
7  both    Slowest days        Average orders         10  
8  both    Slowest days        Average orders         10  
9  both    Slowest days        Average orders         10

In [16]:
# Creating a loyalty flag for existing customers using the loc() function
df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df.loc[(df['max_order'] <= 40) & (df['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df.loc[df['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [12]:
# Printing frequency of the new 'loyalty_flag'
df['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [14]:
# Checking output after last execution
df.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   
15    473747        1             3                   3                 12   
16   2254736        1             4                   4                  7   
17   2254736        1             4                   4                  7   
18   2254736        1             4                   4                  7   
19   2254736        1             4                   4                  7   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      7.0         196                  1          0   
1                      7.0       14084                  2          0   
2                      7.0       12427                  3          0   
3                      7.0       26088                  4          0   
4                      7.0       26405                  5          0   
5                     15.0         196                  1          1   
6                     15.0       10258                  2          0   
7                     15.0       12427                  3          1   
8                     15.0       13176                  4          0   
9                     15.0       26088                  5          1   
10                    15.0       13032                  6          0   
11                    21.0         196                  1          1   
12                    21.0       12427                  2          1   
13                    21.0       10258                  3          1   
14                    21.0       25133                  4          0   
15                    21.0       30450                  5          0   
16                    29.0         196                  1          1   
17                    29.0       12427                  2          1   
18                    29.0       10258                  3          1   
19                    29.0       25133                  4          1   

                               product_name  aisle_id  department_id  prices  \
0                                      Soda        77              7     9.0   
1   Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                       Original Beef Jerky        23             19     4.4   
3                Aged White Cheddar Popcorn        23             19     4.7   
4          XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                      Soda        77              7     9.0   
6                                Pistachios       117             19     3.0   
7                       Original Beef Jerky        23             19     4.4   
8                    Bag of Organic Bananas        24              4    10.3   
9                Aged White Cheddar Popcorn        2

5. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [15]:
# Calculating spending habits based on loyalty tag
df.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

- 'New customers' have the highest mean price at 13.29, followed by 'Regular customers' at 12.50 and 'Loyal customers' lastly at 10.39.
- All customers share the same minimum and maximum prices.

6. Create a spending flag for 'Low spender' and 'High spender'

In [17]:
# Using the transform() function for data aggregation
df['avg_product_price'] = df.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\Jason\AppData\Local\Temp\ipykernel_10408\437501255.py:2: FutureWarning: The provided callable <function mean at 0x0000022CF15B3600> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df['avg_product_price'] = df.groupby(['user_id'])['prices'].transform(np.mean)


In [18]:
# Checking output and columns used
df[['user_id', 'avg_product_price']].head(10)

user_id  avg_product_price
0        1           6.367797
1        1           6.367797
2        1           6.367797
3        1           6.367797
4        1           6.367797
5        1           6.367797
6        1           6.367797
7        1           6.367797
8        1           6.367797
9        1           6.367797

In [25]:
# Creating 'Low spender' and 'High spender' flags using loc function
df.loc[df['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'
df.loc[df['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [20]:
# Printing frequency of the new 'low spender' and 'High spender' flags
df['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [22]:
# Using the head.() function to review the output and columns
df[['user_id', 'avg_product_price', 'spending_flag']].head(20)

user_id  avg_product_price spending_flag
0         1           6.367797   Low spender
1         1           6.367797   Low spender
2         1           6.367797   Low spender
3         1           6.367797   Low spender
4         1           6.367797   Low spender
5         1           6.367797   Low spender
6         1           6.367797   Low spender
7         1           6.367797   Low spender
8         1           6.367797   Low spender
9         1           6.367797   Low spender
10        1           6.367797   Low spender
11        1           6.367797   Low spender
12        1           6.367797   Low spender
13        1           6.367797   Low spender
14        1           6.367797   Low spender
15        1           6.367797   Low spender
16        1           6.367797   Low spender
17        1           6.367797   Low spender
18        1           6.367797   Low spender
19        1           6.367797   Low spender

7. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.

In [23]:
# Using the transform() function for data aggregation
df['median_days_between_orders'] = df.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\Jason\AppData\Local\Temp\ipykernel_10408\2098855144.py:2: FutureWarning: The provided callable <function median at 0x0000022CF1706B60> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df['median_days_between_orders'] = df.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [24]:
# Using the head.() function to review the output and columns
df[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(20)

user_id  days_since_prior_order  median_days_between_orders
0         1                     7.0                        20.0
1         1                     7.0                        20.0
2         1                     7.0                        20.0
3         1                     7.0                        20.0
4         1                     7.0                        20.0
5         1                    15.0                        20.0
6         1                    15.0                        20.0
7         1                    15.0                        20.0
8         1                    15.0                        20.0
9         1                    15.0                        20.0
10        1                    15.0                        20.0
11        1                    21.0                        20.0
12        1                    21.0                        20.0
13        1                    21.0                        20.0
14        1                    21.0                        20.0
15        1                    21.0                        20.0
16        1                    29.0                        20.0
17        1                    29.0                        20.0
18        1                    29.0                        20.0
19        1                    29.0                        20.0

In [27]:
# Creating 'Order frequency' flags using loc function
df.loc[df['median_days_between_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
df.loc[(df['median_days_between_orders'] <= 20) & (df['median_days_between_orders'] > 10), 'order_frequency_flag'] = 'Regular customer'
df.loc[df['median_days_between_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [28]:
# Printing frequency of the 'Order frequency' flags
df['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        22790523
Regular customer          6927608
Non-frequent customer     2686728
Name: count, dtype: int64

In [29]:
# Using the head.() function to review the output and columns
df[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency_flag']].head(20)

user_id  days_since_prior_order  median_days_between_orders  \
0         1                     7.0                        20.0   
1         1                     7.0                        20.0   
2         1                     7.0                        20.0   
3         1                     7.0                        20.0   
4         1                     7.0                        20.0   
5         1                    15.0                        20.0   
6         1                    15.0                        20.0   
7         1                    15.0                        20.0   
8         1                    15.0                        20.0   
9         1                    15.0                        20.0   
10        1                    15.0                        20.0   
11        1                    21.0                        20.0   
12        1                    21.0                        20.0   
13        1                    21.0                        20.0   
14        1                    21.0                        20.0   
15        1                    21.0                        20.0   
16        1                    29.0                        20.0   
17        1                    29.0                        20.0   
18        1                    29.0                        20.0   
19        1                    29.0                        20.0   

   order_frequency_flag  
0      Regular customer  
1      Regular customer  
2      Regular customer  
3      Regular customer  
4      Regular customer  
5      Regular customer  
6      Regular customer  
7      Regular customer  
8      Regular customer  
9      Regular customer  
10     Regular customer  
11     Regular customer  
12     Regular customer  
13     Regular customer  
14     Regular customer  
15     Regular customer  
16     Regular customer  
17     Regular customer  
18     Regular customer  
19     Regular customer

9. Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [30]:
# Exporting df
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_prods_merge_derived.pkl'))